### PR TITLE
Orgs Module For Admin Settings

### PR DESCRIPTION
**Pull request by NileshPant1999**

### CODE REVIEW

File: apps/bamboohr/views.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+from workato_connector.workato import Workato
+
+
+class FyleConnection(viewsets.ViewSet):
+    """
+    Api Call to make Fyle Connection in workato
+    """
+
+    authentication_classes = []
```
Comment (NileshPant1999) [inline apps/bamboohr/views.py:15]: will remove this from everywhere before merging, it was just here for test purpose

File: apps/bamboohr/views.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    permission_classes = []
+
+    def post(self, request, *args, **kwargs):
+
+        connector = Workato()
+        managed_user_id=request.data['managed_user_id']
+        connections = connector.connections.get(managed_user_id=managed_user_id)
+        connection_id = connections['result'][0]['id']
```
Comment (NileshPant1999) [inline apps/bamboohr/views.py:23]: need to change this, will get it from frontend

File: apps/bamboohr/urls.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from django.urls import path
+
+from .views import  FyleConnection, ReadyView, RecipeView
+
+urlpatterns = [
+    path('connect_fyle/', FyleConnection.as_view({'post': 'post'}), name='fyle-connection'),
```
Comment (Shwetabhk) [inline apps/bamboohr/urls.py:6]: Move this to orgs

File: apps/bamboohr/urls.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+    path('recipes/', RecipeView.as_view({'get': 'get'}, name='recipes')),
+    path('recipes/<int:recipe_id>/', RecipeView.as_view({'get': 'get_by_id'}), name='recipe-by-id')
```
Comment (Shwetabhk) [inline apps/bamboohr/urls.py:8]: This endpoint is not required

File: apps/bamboohr/views.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+from requests import Response
+
+from rest_framework.response import Response
+from rest_framework.views import status
+from rest_framework import viewsets
+
+from workato_connector.workato import Workato
+
+
+class FyleConnection(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/bamboohr/views.py:10]: Please use generics

Comment (Shwetabhk) [inline apps/bamboohr/views.py:10]: move to orgs and use generics for all views

File: apps/bamboohr/views.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+from requests import Response
+
+from rest_framework.response import Response
+from rest_framework.views import status
+from rest_framework import viewsets
+
+from workato_connector.workato import Workato
```
Comment (Shwetabhk) [inline apps/bamboohr/views.py:7]: remove _connector
```suggestion
from workato import Workato
```

File: apps/bamboohr/views.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    Api Call to make Fyle Connection in workato
+    """
+
+    authentication_classes = []
+    permission_classes = []
+
+    def post(self, request, *args, **kwargs):
+
+        connector = Workato()
+        managed_user_id=request.data['managed_user_id']
```
Comment (Shwetabhk) [inline apps/bamboohr/views.py:21]: check spacing

File: apps/bamboohr/views.py (hunk 5 above)
Comment (Shwetabhk) [inline apps/bamboohr/views.py:10]:
```suggestion
class FyleConnection(generics.RetrieveUpdateView):
```

File: apps/bamboohr/views.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+class FyleConnection(viewsets.ViewSet):
+    """
+    Api Call to make Fyle Connection in workato
+    """
+
+    authentication_classes = []
+    permission_classes = []
+
+    def post(self, request, *args, **kwargs):
```
Comment (Shwetabhk) [inline apps/bamboohr/views.py:18]: move all logic to serializer

File: apps/bamboohr/views.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+                status=status.HTTP_400_BAD_REQUEST
+            )    
+
+        return Response(
+           connection,
+           status=status.HTTP_200_OK
+        )
+
+
+class RecipeView(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/bamboohr/views.py:46]:
```suggestion
class RecipeView(generics.ListAPIView):
```

File: apps/bamboohr/views.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    API Call to Get Fyle Recipe
+    """
+
+    authentication_classes = []
+    permission_classes = []
+
+
+    def get(self, request, *args, **kwargs):
```
Comment (Shwetabhk) [inline apps/bamboohr/views.py:55]: move get function definition inside serializer

File: apps/bamboohr/views.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        connector = Workato()
+        managed_user_id=request.data['managed_user_id']
+        recipes = connector.recipes.get(managed_user_id=managed_user_id)
+
+        return Response(
+           recipes,
+           status=status.HTTP_200_OK
+        )
+
+    def get_by_id(self, request, *args, **kwargs):
```
Comment (Shwetabhk) [inline apps/bamboohr/views.py:65]: remove this not required

File: apps/orgs/helpers.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
+import json
+from typing import Dict
+
+import requests
+
+from django.conf import settings
+
+
+def post_request(url: str, body: Dict, access_token: str) -> Dict:
```
Comment (Shwetabhk) [inline apps/orgs/helpers.py:9]: can move it to admin_settings.helpers

Comment (Shwetabhk) [inline apps/orgs/helpers.py:9]:
```suggestion
def post_request(url: str, body: Dict, header: dict) -> Dict:
```

File: apps/orgs/models.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class Org(models.Model):
+    """
+    Org Model
+    """
+
+    id = models.AutoField(primary_key=True, help_text='Unique Id to indentify a Org')
+    name = models.CharField(max_length=255, help_text='Name of the org')
+    user = models.ManyToManyField(User, help_text='Reference to users table')
```
Comment (Shwetabhk) [inline apps/orgs/models.py:16]: check if we need to put on_delete property here

File: apps/orgs/models.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    fyle_org_id = models.CharField(max_length=255, help_text='org id', unique=True)
```
Comment (Shwetabhk) [inline apps/orgs/models.py:17]: org_id

File: apps/orgs/models.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    managed_user_id = models.CharField(max_length=255, null=True, help_text='Managed User Id')
+    cluster_domain = models.CharField(max_length=255, help_text='Fyle cluster domain')
+    is_bamboo_connector = models.BooleanField(null=True, help_text='Is Bamboo Hr Connected')
```
Comment (Shwetabhk) [inline apps/orgs/models.py:20]:
```suggestion
    is_bamboo_connected = models.BooleanField(null=True, help_text='Is Bamboo Hr Connected')
```

Comment (Shwetabhk) [inline apps/orgs/models.py:20]:
```suggestion
    is_bamboo_connected = models.BooleanField(default=False, help_text='Is Bamboo Hr Connected')
```

File: apps/orgs/urls.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+org_app_path = [
+    path('', OrgsView.as_view({'get': 'get', 'post': 'post'}), name='orgs'),
+    path('<int:org_id>/', OrgsView.as_view({'get': 'get_by_id'}), name='org-by-id'),
+    path('<int:org_id>/connect_workato', OrgsView.as_view({'post': 'post'}), name='connect workato'),
+    path('<int:org_id>/managed_users/', ManagedUser.as_view({'get': 'get', 'post': 'post'}), name='managed-users'),
+    path('ready/', ReadyView.as_view({'get': 'get'}), name='ready'),
+]
+
+other_app_path = [
```
Comment (Shwetabhk) [inline apps/orgs/urls.py:13]: integration_paths

File: apps/orgs/urls.py [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    path('<int:org_id>/bamboohr/', include('apps.bamboohr.urls'))
+]
+
+urlpatterns = []
```
Comment (Shwetabhk) [inline apps/orgs/urls.py:17]:
```suggestion
urlpatterns = [
    **org_app_path,
    **other_app_path
]
```

File: apps/orgs/views.py [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+from fyle_rest_auth.models import AuthToken
+from fyle_rest_auth.helpers import get_fyle_admin
+from apps.orgs.helpers import get_cluster_domain
+
+from apps.orgs.serializers import OrgSerializer
+from apps.orgs.models import FyleCredential, Org, User
+from workato_connector.workato import Workato
+
+
+class ReadyView(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/orgs/views.py:19]: RetrieveAPIView

File: apps/orgs/views.py [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+        Org.objects.first()
+
+        return Response(
+            data={
+                'message': 'Ready'
+            },
+            status=status.HTTP_200_OK
+        )
+
+class OrgsView(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/orgs/views.py:40]: use generics

Comment (Shwetabhk) [reply apps/orgs/views.py:40]: RetrieveUpdate

File: apps/orgs/views.py [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+    """
+    Fyle Org
+    """
+
+    permission_classes = [IsAuthenticated]
+
+    def post(self, request):
```
Comment (Shwetabhk) [inline apps/orgs/views.py:47]: move all logic to serializer

File: apps/orgs/views.py [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+        except Org.DoesNotExist:
+            return Response(
+                data={
+                    'message': 'Org with this id does not exist'
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+
+class ManagedUser(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/orgs/views.py:112]: same comments as above

File: apps/orgs/views.py [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+        return Response(
+            managed_user,
+            status=status.HTTP_200_OK
+        )
+    
+    def get(self, request, *args, **kwargs):
+
+        connector = Workato()
+
+        managed_users = connector.managed_users.get()
```
Comment (Shwetabhk) [inline apps/orgs/views.py:139]: should filter based on org id and get the specific managed user

File: workato_connector/apis/api_base.py [hunk 23]
```diff
@@ -39,6 +39,7 @@ def _put_request(self, url: str, data: dict = None, action: string = None) -> re
         if action:
             response = requests.put('{}{}/{}'.format(API_URL, url, action), headers=API_HEADERS)
         else:
+            print('data', data)
```
Comment (Shwetabhk) [inline workato_connector/apis/api_base.py:42]: remove print

File: apps/orgs/models.py (hunk 13 above)
Comment (NileshPant1999) [reply apps/orgs/models.py:16]: no need we are not doing anywhere else

**Review by Shwetabhk — APPROVED** [review]

**Pull request merged**